src/theory/model_manager.cpp

```diff
@@ -0,0 +1,177 @@
+/*********************                                                        */
+/*! \file model_manager.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Management of a distributed approach for equality engines over
+ ** all theories.
+ **/
+
+#include "theory/model_manager.h"
+
+#include "options/theory_options.h"
+#include "theory/quantifiers_engine.h"
+#include "theory/theory_engine.h"
+
+namespace CVC4 {
+namespace theory {
+
+ModelManager::ModelManager(TheoryEngine& te)
+    : d_te(te),
+      d_logicInfo(te.getLogicInfo()),
+      d_model(nullptr),
+      d_modelBuilder(nullptr),
+      d_modelBuilt(false),
+      d_modelBuiltSuccess(false)
+{
+}
+
+ModelManager::~ModelManager() {}
+
+void ModelManager::finishInit()
+{
+  // construct the model
+  const LogicInfo& logicInfo = d_te.getLogicInfo();
+  // Initialize the model and model builder.
+  if (logicInfo.isQuantified())
+  {
+    QuantifiersEngine* qe = d_te.getQuantifiersEngine();
+    Assert(qe != nullptr);
+    d_modelBuilder = qe->getModelBuilder();
+    d_model = qe->getModel();
+  }
+  else
+  {
+    context::Context* u = d_te.getUserContext();
+    d_alocModel.reset(
+        new TheoryModel(u, "DefaultModel", options::assignFunctionValues()));
+    d_model = d_alocModel.get();
+  }
+
+  // make the default builder, e.g. in the case that the quantifiers engine does
+  // not have a model builder
+  if (d_modelBuilder == nullptr)
+  {
+    d_alocModelBuilder.reset(new TheoryEngineModelBuilder(&d_te));
+    d_modelBuilder = d_alocModelBuilder.get();
+  }
+  // notice that the equality engine of the model has yet to be assigned.
+}
+
+void ModelManager::resetModel()
+{
+  d_modelBuilt = false;
+  d_modelBuiltSuccess = false;
+  // Reset basic information on the model object
+  d_model->reset();
+}
+
+bool ModelManager::buildModel()
+{
+  if (d_modelBuilt)
+  {
+    // already computed
+    return d_modelBuiltSuccess;
+  }
+  // reset the flags now
+  d_modelBuilt = true;
+  d_modelBuiltSuccess = false;
+
+  // prepare the model, which is specific to the manager
+  if (!prepareModel())
+  {
+    Trace("model-builder") << "ModelManager: fail prepare model" << std::endl;
+    return false;
+  }
+
+  // now, finish building the model
+  d_modelBuiltSuccess = finishBuildModel();
```

Comment:
The method `finishBuildModel()` seems to essentially be a one-liner. Maybe inline that function here?

Reply:
It's actually plan to be called manually by the model-based combination method.